File: red/src/operator.ts
import * as t from "./tools.ts";
import type { Obj } from "./tools.ts";
export const unchanged = (cr: Obj | null, e: Obj, key: string) =>
  !!cr &&
  cr.metadata?.uid === e.resourceUID &&
  cr.metadata?.generation === e[key];
export function requireUnchanged(cr: Obj | null, e: Obj, key: string): Obj {
  if (!unchanged(cr, e, key))
    throw new t.Fatal("Desired state changed during operation");
  return cr!;
}
export async function check(o: Obj) {
  await t.waitReady(o);
  console.log("check", await t.healthyProbe(o));
  const directory = "/data/work/" + o.profile + "/failures";
  const files = await t.kubectl(
    o,
    [
      "exec",
      "deployment/" + t.CONTROLLER,
      "-n",
      o.namespace,
      "--",
      "sh",
      "-c",
      `ls -1 ${directory} 2>/dev/null || true`,
    ],
    { requestTimeout: false, timeout: 60 },
  );
  console.log(
    "failures retained:",
    files.trim() ? files.trim().split("\n").length : 0,
  );
}
export async function rehearse(o: Obj) {
  const original = await t.waitReady(o),
    e: Obj = {
      startedAt: t.now(),
      resourceUID: original.metadata.uid,
      passed: false,
    };
  const save = () => t.evidence(o, "backup-rehearsal", e);
  save();
  const suspended = await t.patch(o, original, [
    { op: "add", path: "/spec/suspend", value: true },
  ]);
  e.suspendedGeneration = suspended.metadata.generation;
  save();
  let failure: any,
    uncertain = false;
  try {
    await t.waitFor(
      "acknowledged suspension",
      async () => {
        const cr = requireUnchanged(
          await t.getResource(o),
          e,
          "suspendedGeneration",
        );
        return t.acknowledged(cr) ? cr : null;
      },
      7800,
      3,
    );
    e.suspensionAcknowledgedAt = t.now();
    save();
    let result: Obj;
    try {
      result = await t.probe(o, "rehearse");
    } catch (error) {
      uncertain = true;
      throw error;
    }
    e.result = result;
    if (result.rehearsalPassed !== true)
      throw new Error("Backup rehearsal failed");
    e.passed = true;
  } catch (error) {
    failure = error;
  }
  try {
    const current = await t.getResource(o);
    if (uncertain) e.resumeBlocked = "Remote execution completion is uncertain";
    else if (!unchanged(current, e, "suspendedGeneration"))
      e.resumeBlocked = "Resource changed during rehearsal";
    else {
      const resumed = await t.patch(o, current, [
        { op: "add", path: "/spec/suspend", value: false },
      ]);
      e.resumedGeneration = resumed.metadata.generation;
      e.resumedAt = t.now();
    }
  } catch (error) {
    e.resumeBlocked = "Could not re-read or resume resource: " + String(error);
  }
  const path = save();
  if (failure || e.resumeBlocked)
    throw new Error(
      String(failure ?? e.resumeBlocked) +
        (e.resumeBlocked
          ? "; " +
            e.resumeBlocked +
            "; resource left suspended; verify no workflow runs before resuming"
          : "") +
        "; evidence: " +
        path,
    );
  await t.waitReady(o, 2700);
}
const uuid = () => crypto.randomUUID().replaceAll("-", "");
export async function drill(o: Obj) {
  const cr = await t.waitReady(o),
    before = await t.healthyProbe(o),
    token = o["do-token"],
    cluster = o["doks-cluster-id"],
    workers = new Set<string>();
  if (cluster) {
    const result = await t.digitalocean(
      token,
      "get",
      "kubernetes/clusters/" + cluster,
    );
    for (const pool of result!.kubernetes_cluster.node_pools)
      for (const node of pool.nodes) workers.add(String(node.droplet_id));
  }
  const droplet = await t.digitalocean(
    token,
    "get",
    "droplets/" + before.providerId,
  );
  t.ownedDroplet(droplet?.droplet, before, o.profile, workers);
  const key = "colors:self-heal:" + uuid(),
    value = uuid();
  if ((await t.probe(o, "set-marker", key, value)).marker !== value)
    throw new t.Fatal("Authenticated SET/GET failed before disruption");
  const e: Obj = {
    startedAt: t.now(),
    clusterId: cluster ?? null,
    resourceUID: cr.metadata.uid,
    generation: cr.metadata.generation,
    before,
    excludedWorkerIds: [...workers].sort(),
    markerKey: key,
    markerValue: value,
    expectedDataRecovery: false,
    passed: false,
  };
  const save = () => t.evidence(o, "self-healing", e);
  save();
  await t.digitalocean(token, "delete", "droplets/" + before.providerId);
  e.deleteAcceptedAt = t.now();
  save();
  try {
    await t.waitFor(
      "service recovery",
      async () => {
        const current = requireUnchanged(
          await t.getResource(o),
          e,
          "generation",
        );
        if (t.suspended(current) || t.deleting(current))
          throw new t.Fatal("Resource suspended or deleting during recovery");
        const after = await t.probe(o, "health");
        if (
          after.providerId === before.providerId ||
          after.healthy !== true ||
          !t.ready(current)
        )
          return null;
        const live = await t.digitalocean(
          token,
          "get",
          "droplets/" + after.providerId,
        );
        t.ownedDroplet(live?.droplet, after, o.profile, workers);
        if (
          (await t.digitalocean(
            token,
            "get",
            "droplets/" + before.providerId,
          )) !== null
        )
          throw new t.Fatal("Original Droplet still exists");
        const survived = (await t.probe(o, "get-marker", key)).marker === value,
          fresh = uuid(),
          written =
            (await t.probe(o, "set-marker", key, fresh)).marker === fresh;
        Object.assign(e, {
          recoveredAt: t.now(),
          after,
          priorMarkerSurvived: survived,
          authenticatedWriteReadPassed: written,
        });
        if (!written) throw new t.Fatal("Replacement write/read failed");
        return after;
      },
      2400,
      15,
    );
  } catch (error) {
    Object.assign(e, {
      timedOutAt: t.now(),
      lastErrorType: error instanceof t.Fatal ? "fatal" : "timeout",
      lastError: String(error),
    });
    save();
    throw error;
  }
  e.passed = true;
  save();
}
export async function restart(o: Obj) {
  const cr = await t.waitReady(o),
    before = await t.healthyProbe(o),
    namespace = o.namespace,
    deployment = "deployment/" + t.CONTROLLER;
  const doc = await t.kubectl(
    o,
    ["get", "deployment", t.CONTROLLER, "-n", namespace, "-o", "json"],
    { parse: true },
  );
  if (doc.spec.replicas !== 1)
    throw new t.Fatal("Restart requires exactly one controller replica");
  const old = await t.controllerPod(o);
  const e: Obj = {
    startedAt: t.now(),
    before,
    resourceUID: cr.metadata.uid,
    generation: cr.metadata.generation,
    lastReconcileTimeBefore: cr.status?.lastReconcileTime,
    oldPod: old?.metadata.name,
    passed: false,
  };
  const save = () => t.evidence(o, "controller-restart", e);
  save();
  await t.kubectl(o, ["scale", deployment, "-n", namespace, "--replicas=0"]);
  try {
    await t.waitFor(
      "old controller to stop",
      async () =>
        !(
          await t.kubectl(
            o,
            [
              "get",
              "pods",
              "-n",
              namespace,
              "-l",
              "app=" + t.CONTROLLER,
              "-o",
              "json",
            ],
            { parse: true },
          )
        ).items.length,
      7800,
    );
  } catch (error) {
    e.failure =
      "Old controller did not stop; intentionally leaving replicas at zero";
    save();
    throw error;
  }
  e.stoppedAt = t.now();
  save();
  await t.kubectl(o, ["scale", deployment, "-n", namespace, "--replicas=1"]);
  await t.kubectl(
    o,
    ["rollout", "status", deployment, "-n", namespace, "--timeout=600s"],
    { requestTimeout: false, timeout: 630 },
  );
  const pod = await t.controllerUp(o),
    start = pod.status?.startTime;
  if (!start || (old && old.metadata.uid === pod.metadata.uid))
    throw new t.Fatal("Controller pod did not change or has no start time");
  e.newPod = pod.metadata.name;
  e.newPodStartTime = start;
  save();
  const current = await t.waitFor(
      "reconcile by new controller",
      async () => {
        const current = t.active(await t.getResource(o)),
          timestamp = current.status?.lastReconcileTime;
        return t.ready(current) &&
          timestamp &&
          Date.parse(timestamp) > Date.parse(start)
          ? current
          : null;
      },
      1800,
      10,
    ),
    after = await t.probe(o, "health");
  if (after.healthy !== true || before.providerId !== after.providerId)
    throw new t.Fatal("Restart did not preserve healthy infrastructure");
  if (
    before.convergenceRecordModifiedMs == null ||
    before.convergenceRecordModifiedMs !== after.convergenceRecordModifiedMs
  )
    throw new t.Fatal("Restart unexpectedly ran convergence");
  requireUnchanged(current, e, "generation");
  Object.assign(e, {
    passed: true,
    after,
    completedAt: t.now(),
    lastReconcileTimeAfter: current.status.lastReconcileTime,
  });
  save();
}

File: blue/src/package_redis_operator_blue/operator.py
"""Operational checks with guarded suspension and disruption."""

from . import tools as t
from uuid import uuid4
from datetime import datetime


def unchanged(cr, e, key):
    return (
        cr is not None
        and cr["metadata"].get("uid") == e["resourceUID"]
        and cr["metadata"].get("generation") == e[key]
    )


def require_unchanged(cr, e, key):
    if not unchanged(cr, e, key):
        raise t.Fatal("Desired state changed during operation")
    return cr


async def check(opts):
    await t.wait_ready(opts)
    observed = await t.healthy_probe(opts)
    print("check", observed)
    directory = "/data/work/" + opts["profile"] + "/failures"
    files = await t.kubectl(
        opts,
        [
            "exec",
            "deployment/" + t.CONTROLLER,
            "-n",
            opts["namespace"],
            "--",
            "sh",
            "-c",
            f"ls -1 {directory} 2>/dev/null || true",
        ],
        request_timeout=False,
        timeout=60,
    )
    print("failures retained:", len(files.splitlines()))


async def rehearse(opts):
    original = await t.wait_ready(opts)
    evidence = {
        "startedAt": t.now(),
        "resourceUID": original["metadata"]["uid"],
        "passed": False,
    }

    def save():
        return t.evidence(opts, "backup-rehearsal", evidence)

    save()
    suspended = await t.patch(
        opts, original, [{"op": "add", "path": "/spec/suspend", "value": True}]
    )
    evidence["suspendedGeneration"] = suspended["metadata"]["generation"]
    save()
    failure = None
    uncertain = False
    try:

        async def acknowledged():
            current = require_unchanged(
                await t.get_resource(opts), evidence, "suspendedGeneration"
            )
            return current if t.acknowledged(current) else None

        await t.wait_for("acknowledged suspension", acknowledged, 7800, 3)
        evidence["suspensionAcknowledgedAt"] = t.now()
        save()
        try:
            result = await t.probe(opts, "rehearse")
        except Exception:
            uncertain = True
            raise
        evidence["result"] = result
        if result.get("rehearsalPassed") is not True:
            raise RuntimeError("Backup rehearsal failed")
        evidence["passed"] = True
    except Exception as exc:
        failure = exc
    try:
        current = await t.get_resource(opts)
        if uncertain:
            evidence["resumeBlocked"] = "Remote execution completion is uncertain"
        elif not unchanged(current, evidence, "suspendedGeneration"):
            evidence["resumeBlocked"] = "Resource changed during rehearsal"
        else:
            resumed = await t.patch(
                opts, current, [{"op": "add", "path": "/spec/suspend", "value": False}]
            )
            evidence.update(
                resumedGeneration=resumed["metadata"]["generation"], resumedAt=t.now()
            )
    except Exception as exc:
        evidence["resumeBlocked"] = "Could not re-read or resume resource: " + str(exc)
    path = save()
    blocked = evidence.get("resumeBlocked")
    if failure or blocked:
        raise RuntimeError(
            str(failure or blocked)
            + (
                (
                    "; "
                    + blocked
                    + "; resource left suspended; verify no workflow runs before resuming"
                )
                if blocked
                else ""
            )
            + "; evidence: "
            + path
        )
    await t.wait_ready(opts, 2700)


async def drill(opts):
    cr = await t.wait_ready(opts)
    before = await t.healthy_probe(opts)
    token = opts["do-token"]
    cluster = opts.get("doks-cluster-id")
    workers = set()
    if cluster:
        result = await t.digitalocean(token, "get", "kubernetes/clusters/" + cluster)
        workers = {
            str(n["droplet_id"])
            for p in result["kubernetes_cluster"]["node_pools"]
            for n in p["nodes"]
        }
    droplet = await t.digitalocean(token, "get", "droplets/" + before["providerId"])
    t.owned_droplet((droplet or {}).get("droplet"), before, opts["profile"], workers)
    key = "colors:self-heal:" + uuid4().hex
    value = uuid4().hex
    if (await t.probe(opts, "set-marker", key, value)).get("marker") != value:
        raise t.Fatal("Authenticated SET/GET failed before disruption")
    evidence = {
        "startedAt": t.now(),
        "clusterId": cluster,
        "resourceUID": cr["metadata"]["uid"],
        "generation": cr["metadata"]["generation"],
        "before": before,
        "excludedWorkerIds": sorted(workers),
        "markerKey": key,
        "markerValue": value,
        "expectedDataRecovery": False,
        "passed": False,
    }

    def save():
        return t.evidence(opts, "self-healing", evidence)

    save()
    await t.digitalocean(token, "delete", "droplets/" + before["providerId"])
    evidence["deleteAcceptedAt"] = t.now()
    save()

    async def recovered():
        current = require_unchanged(await t.get_resource(opts), evidence, "generation")
        if t.suspended(current) or t.deleting(current):
            raise t.Fatal("Resource suspended or deleting during recovery")
        after = await t.probe(opts, "health")
        if (
            after.get("providerId") == before["providerId"]
            or after.get("healthy") is not True
            or not t.ready(current)
        ):
            return None
        live = await t.digitalocean(token, "get", "droplets/" + after["providerId"])
        t.owned_droplet((live or {}).get("droplet"), after, opts["profile"], workers)
        if (
            await t.digitalocean(token, "get", "droplets/" + before["providerId"])
            is not None
        ):
            raise t.Fatal("Original Droplet still exists")
        survived = (await t.probe(opts, "get-marker", key)).get("marker") == value
        fresh = uuid4().hex
        written = (await t.probe(opts, "set-marker", key, fresh)).get("marker") == fresh
        evidence.update(
            recoveredAt=t.now(),
            after=after,
            priorMarkerSurvived=survived,
            authenticatedWriteReadPassed=written,
        )
        if not written:
            raise t.Fatal("Replacement write/read failed")
        return after

    try:
        await t.wait_for("service recovery", recovered, 2400, 15)
    except Exception as exc:
        evidence.update(
            timedOutAt=t.now(),
            lastErrorType="fatal" if isinstance(exc, t.Fatal) else "timeout",
            lastError=str(exc),
        )
        save()
        raise
    evidence["passed"] = True
    save()


async def restart(opts):
    cr = await t.wait_ready(opts)
    before = await t.healthy_probe(opts)
    namespace = opts["namespace"]
    deployment = "deployment/" + t.CONTROLLER
    doc = await t.kubectl(
        opts,
        ["get", "deployment", t.CONTROLLER, "-n", namespace, "-o", "json"],
        parse=True,
    )
    if doc["spec"].get("replicas") != 1:
        raise t.Fatal("Restart requires exactly one controller replica")
    old = await t.controller_pod(opts)
    evidence = {
        "startedAt": t.now(),
        "before": before,
        "resourceUID": cr["metadata"]["uid"],
        "generation": cr["metadata"]["generation"],
        "lastReconcileTimeBefore": cr.get("status", {}).get("lastReconcileTime"),
        "oldPod": (old or {}).get("metadata", {}).get("name"),
        "passed": False,
    }

    def save():
        return t.evidence(opts, "controller-restart", evidence)

    save()
    await t.kubectl(opts, ["scale", deployment, "-n", namespace, "--replicas=0"])

    async def stopped():
        return not (
            await t.kubectl(
                opts,
                [
                    "get",
                    "pods",
                    "-n",
                    namespace,
                    "-l",
                    "app=" + t.CONTROLLER,
                    "-o",
                    "json",
                ],
                parse=True,
            )
        ).get("items")

    try:
        await t.wait_for("old controller to stop", stopped, 7800)
    except Exception:
        evidence["failure"] = (
            "Old controller did not stop; intentionally leaving replicas at zero"
        )
        save()
        raise
    evidence["stoppedAt"] = t.now()
    save()
    await t.kubectl(opts, ["scale", deployment, "-n", namespace, "--replicas=1"])
    await t.kubectl(
        opts,
        ["rollout", "status", deployment, "-n", namespace, "--timeout=600s"],
        request_timeout=False,
        timeout=630,
    )
    pod = await t.controller_up(opts)
    start = pod["status"].get("startTime")
    if not start or old and old["metadata"]["uid"] == pod["metadata"]["uid"]:
        raise t.Fatal("Controller pod did not change or has no start time")
    evidence.update(newPod=pod["metadata"]["name"], newPodStartTime=start)
    save()

    async def reconciled():
        current = t.active(await t.get_resource(opts))
        timestamp = current.get("status", {}).get("lastReconcileTime")
        return (
            current
            if t.ready(current)
            and timestamp
            and datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
            > datetime.fromisoformat(start.replace("Z", "+00:00"))
            else None
        )

    current = await t.wait_for("reconcile by new controller", reconciled, 1800, 10)
    after = await t.probe(opts, "health")
    if after.get("healthy") is not True or before["providerId"] != after.get(
        "providerId"
    ):
        raise t.Fatal("Restart did not preserve healthy infrastructure")
    if before.get("convergenceRecordModifiedMs") is None or before[
        "convergenceRecordModifiedMs"
    ] != after.get("convergenceRecordModifiedMs"):
        raise t.Fatal("Restart unexpectedly ran convergence")
    require_unchanged(current, evidence, "generation")
    evidence.update(
        passed=True,
        after=after,
        completedAt=t.now(),
        lastReconcileTimeAfter=current["status"]["lastReconcileTime"],
    )
    save()
